runtime/hash.c

```diff
@@ -237,12 +230,27 @@ CAMLprim value caml_hash(value count, value limit, value seed, value obj)
         h = caml_hash_mix_uint32(h, Infix_offset_val(v));
         v = v - Infix_offset_val(v);
         goto again;
+      case Closure_tag:
+        /* !! something complicated */
```

Comment:
The `Closure_tag` case is handled  correctly in the current trunk (in no-naked-pointers mode), https://github.com/ocaml/ocaml/blob/35af4cddfd31129391f904167236270a004037f8/runtime/hash.c#L264-L282